modules/system/database/migrations/2021_06_25_143510_Db_Attachment_ID_To_int.php

```diff
@@ -0,0 +1,72 @@
+<?php
+
+use Illuminate\Support\Facades\DB;
+use Winter\Storm\Database\Schema\Blueprint;
+use Winter\Storm\Database\Updates\Migration;
+use Winter\Storm\Support\Facades\Schema;
+
+class DbAttachmentIdToInt extends Migration
+{
+    /**
+     * Run the migrations.
+     *
+     * @return void
+     */
+    public function up()
+    {
+        Schema::table('system_files', function (Blueprint $table) {
+            // Drop old indexes
+            $table->dropIndex(['attachment_id']);
+            $table->dropIndex(['attachment_type']);
+            $table->dropIndex(['field']);
+
+            $table->index([
+                'field',
+                'attachment_type',
+                'attachment_id',
+            ]);
+        });
+
+        // Convert attachment_id to bigint if the attachment_id column only contains integer values.
+        $convert = true;
+        DB::table('system_files')
+            ->select(['id', 'attachment_id'])
+            ->chunkById(1000, function ($attachmentIds) use (&$convert) {
+                $files = $attachmentIds->filter(function ($file) {
+                    return ((string) $file->attachment_id === (string) intval($file->attachment_id));
+                });
+
+                if ($files->isNotEmpty()) {
+                    return $convert = false;
+                }
+            });
+
+        if ($convert) {
+            Schema::table('system_files', function (Blueprint $table) {
+                $table->unsignedBigInteger('attachment_id')->change();
+            });
+        }
```

Comment:
Suggestion on how it would be possible to know it conversion is possible:
The function `CAST AS` is valid in: [MySQL](https://dev.mysql.com/doc/refman/5.7/en/cast-functions.html#function_cast), [SQLite](https://sqlite.org/lang_expr.html#castexpr), [PosgreSQL](https://www.postgresql.org/docs/9.2/typeconv-oper.html) and [MSSQL](https://docs.microsoft.com/en-us/sql/t-sql/functions/cast-and-convert-transact-sql?view=sql-server-ver15) and probably many other DBMS.

The same way you are doing the double type-casting comparison on PHP, this should work **(testing needed)**:
```php
try {
    $possible =
        DB::table('system_files')
            ->selectRaw('
                SELECT (varchar_ids.total = numeric_ids.total) AS possible
                FROM (
                    SELECT COUNT(*) AS total
                    FROM system_files
                    ) AS varchar_ids
                JOIN (
                    SELECT COUNT(*) AS total
                    FROM system_files
                    WHERE CAST(CAST(attachment_id AS INTEGER) AS VARCHAR) = attachment_id
                    ) AS numeric_ids
                ON true
            ')->value('possible');
} catch (\Exception $e) {
    // Some DB raise an exception when typecast is not possible
    // Abort as this mean conversion is not possible
    return;
}

if ($possible) {
    Schema::table('system_files', function (Blueprint $table) {
        $table->unsignedBigInteger('attachment_id')->change();
    });

    Schema::table('system_files', function (Blueprint $table) {
        // Drop old indexes
        $table->dropIndex('attachment_id');
        $table->dropIndex('attachment_type');
        $table->dropIndex('field');

        $table->index([
            'field',
            'attachment_type',
            'attachment_id',
        ]);
    });
}
```

Reply:
Hey, thank you for attempting to solve this via SQL but this query does not execute for me when running directly on the database (without going through Laravel/PHP).

Here's the error I get back (MariaDB 10.3):
```bash
[42000][1064] (conn=30) You have an error in your SQL syntax; check the manual that corresponds to your MariaDB server version for the right syntax to use near ') = attachment_id [42000][1064] You have an error in your SQL syntax; check the manual that corresponds to your MariaDB server version for the right syntax to use near ') = attachment_id ) AS numeric_ids ON true' at line 9
```

Reply:
I thought the same after your answer but using [a syntax checker](https://en.rakko.tools/tools/36/), even `CHAR(n)` doesn't valid, the problem seems to be the double typecasting

Reply:
I believe the length of the CHAR cast is optional based on my understanding when reading the MySQL docs.

Here's some working code that executes a completely custom/raw query inside of Laravel:
```php
$possible = DB::selectOne(DB::raw('
                SELECT (varchar_ids.total = numeric_ids.total) AS possible
                    FROM (
                        SELECT COUNT(*) AS total
                        FROM system_files
                        ) AS varchar_ids
                    JOIN (
                        SELECT COUNT(*) AS total
                        FROM system_files
                        WHERE CAST(CAST(attachment_id AS INTEGER) AS CHAR) = attachment_id
                        ) AS numeric_ids
                    ON true
            '))->possible;
```

I'll try and do some testing with Postgres and SQLite. I just don't know how to setup a MSSQL database since I've never used it before, so I'm not able to test this there.

Reply:
I'm not sure we need to support MSSQL as I can't find any mention of it into the documentation (ping @LukeTowers).

Also `CHAR` in your last comment will return only the first character on some DMBS (just tested on PostgreSQL), you need to specify a length inside parenthesis. System default is 191 since laravel uses utf8mb4 character set but a higher value won't break anything so we are safe with: `CHAR(300)`

Reply:
@RomainMazB SQLServer is supported by Laravel so yes, we need to support MSSQL.

Reply:
Ok, so we should use the conditions I've made into [my comment](https://github.com/wintercms/winter/pull/216#discussion_r671536613).
Equivalent types for type casting are:
- INT and CHAR on MSSQL ([Basic documentation example similar to us using both INT and CHAR](https://docs.microsoft.com/en-us/sql/t-sql/functions/cast-and-convert-transact-sql?view=sql-server-ver15#e-using-cast-with-the-like-clause))
- SIGNED INTEGER and CHAR on MySQL/MariaDB ([Types cast MySQL documentation](https://dev.mysql.com/doc/refman/5.7/en/cast-functions.html#function_cast) / [Similar on MariaDB](https://mariadb.com/kb/en/cast/))
- INTEGER and VARCHAR on SQLite ([Types cast SQLite documentation](https://sqlite.org/datatype3.html#affinity_name_examples))
- INTEGER and CHAR on PostgreSQL ([PostgreSQL datatypes documentation](https://www.postgresql.org/docs/9.5/datatype-numeric.html))
```php
$connection = DB::getDefaultConnection();

$driver = config("database.connections.{$connection}.driver");

$textType = 'CHAR';
$integerType = 'INTEGER';

if ($driver === 'mssql') {
  $integerType = 'INT';
} elseif ($driver === 'mysql') {
  $integerType = 'SIGNED INTEGER';
} elseif ($driver === 'sqlite') {
  $textType = 'VARCHAR';
}

$possible =
  DB::selectOne(DB::raw("
      SELECT (varchar_ids.total = numeric_ids.total) AS possible
          FROM (
              SELECT COUNT(*) AS total
              FROM system_files
              ) AS varchar_ids
          JOIN (
              SELECT COUNT(*) AS total
              FROM system_files
              WHERE CAST(CAST(attachment_id AS $integerType) AS $textType(300)) = attachment_id
              ) AS numeric_ids
          ON true
  "))->possible;
```

Reply:
@RomainMazB @ericp-mrel I'm still leaning towards us keeping the type check in PHP using the method already in this PR. This query is getting too complex for a query that's going to be run once (on migration) and that's it. I don't think there's any expectation by users that migrations have to be "fast", so I don't mind if this conversion takes a couple of seconds.